### ISSUE TITLE
DataLoader with IterativeDataset throws an error when providing a BatchSampler

### ISSUE THREAD

**Issue by tchaton** [pytorch/pytorch#61281]
## To Reproduce

```py
DataLoader(iterative_dataset, batch_sampler=...)
```

## Expected behavior

The DataLoader shouldn't raise an error and let control for the user to properly modify the batch sampler if needed. This would be useful for a MetaLearning application.

## Environment

Please copy and paste the output from our
[environment collection script](https://raw.githubusercontent.com/pytorch/pytorch/master/torch/utils/collect_env.py)
(or fill out the checklist below manually).

You can get the script and run it with:
```
wget https://raw.githubusercontent.com/pytorch/pytorch/master/torch/utils/collect_env.py
# For security purposes, please check the contents of collect_env.py before running it.
python collect_env.py
```

 - PyTorch Version (e.g., 1.0):
 - OS (e.g., Linux):
 - How you installed PyTorch (`conda`, `pip`, source):
 - Build command you used (if compiling from source):
 - Python version:
 - CUDA/cuDNN version:
 - GPU models and configuration:
 - Any other relevant information:

**Comment by heitorschueroff** [comment]
related https://github.com/huggingface/transformers/issues/5829

**Comment by ejguan** [comment]
It's intentional to error out IterableDataset with BatchSampler. The native behavior of IterableDataset should yield data one by one in a sequence. But BatchSampler requires to random access (`__getitem__`) element in Dataset, which is not provided from IterableDataset.

It would be great if you could elaborate your use case.

**Comment by ejguan** [comment]
Please check the comment here:
https://github.com/pytorch/pytorch/blob/91c076eadc9f61828f66fb702b3f99109bf97915/torch/utils/data/dataloader.py#L202-L226

**Comment by albertz** [comment]
I just want to comment that we have always used this in the past in our (mostly TF-based so far) framework [RETURNN](https://github.com/rwth-i6/returnn/). Specifically, our batching logic is like this (pseudo code):
```python
batch = Batch()
for seq in dataset:
    new_num_seqs = batch.num_seqs() + 1
    new_max_sizes = {k: max(batch.max_size(k), len(seq[k])) for k in data_keys}
    new_batch_sizes = {k: new_num_seqs * new_max_sizes[k] for k in data_keys}
    if (
        all(new_batch_sizes[k] <= batch_size_limit[k] for k in data_keys)
        and new_num_seqs <= batch_num_seqs_limit
    ):
        batch.add(seq)
    else:
        yield batch
        batch = Batch()
```
We configure upper limits on the number of sequences in the batch (`batch_num_seqs_limit`) and also on the total number of elements (frames) per data key including the padding (`batch_size_limit`).

(Another trick we use is to sort buckets of sequences by length, after global shuffling, to reduce the amount of padding. See our papers for details, see [here](https://github.com/rwth-i6/returnn).)

Yet another way to have some similar effect is to bucket sequences together of similar sequence lengths. This is for example implemented in TensorFlow [`tf.data.Dataset.bucket_by_sequence_length`](https://www.tensorflow.org/api_docs/python/tf/data/Dataset#bucket_by_sequence_length).

Both approaches work on-the-fly, for iterable datasets, as you see. But both approaches need a custom batch sampler.

**Comment by ejguan** [comment]
@albertz I think torchdata does offer the solution to solve custom batching.
Please check out our [repo](https://github.com/pytorch/data) and [online doc](https://pytorch.org/data/beta/index.html) to see if it meets your need.

**Comment by albertz** [comment]
> I think torchdata does offer the solution to solve custom batching.
> Please check out our [repo](https://github.com/pytorch/data) and [online doc](https://pytorch.org/data/beta/index.html) to see if it meets your need.

Oh interesting, I was not aware of that. That looks very related indeed.

However, for what I described, there are similar functions, but none would really behave exactly this way.

* For the first variant with upper limits for `batch_num_seqs_limit`/`batch_size_limit`, the `MaxTokenBucketizer` looks somewhat related, but not exactly:
   - The algorithm I presented does not need any buffer. I think it's not exactly the same.
   - One actual big difference: `batch_size_limit` is per data key. However, `MaxTokenBucketizer` only has `max_token_count` as a single int, with a single `len_fn`, so it can only consider one of the data keys, and it would ignore the others. I don't really understand this. E.g. for translation, while source/target sequence lengths correlate somewhat, they don't strictly do, and you likely want to have a `max_token_count` separate for source/target.
* Sort buckets of sequences by length, after global shuffling: It sounds as if `BucketBatcher` would do that. However, `BucketBatcher` does more, it also does the batching. I wonder, the motivation of TorchData was to decouple the functionality, however, `BucketBatcher` does two things, which could be decoupled but which are not, namely: 1. sort buckets, and then 2. batching.
* Bucket sequences together of similar sequence lengths, just exactly as [`tf.data.Dataset.bucket_by_sequence_length`](https://www.tensorflow.org/api_docs/python/tf/data/Dataset#bucket_by_sequence_length): Again `BucketBatcher` is similar to that, but not exactly. TF `bucket_by_sequence_length` allows to directly specify `bucket_boundaries` and `bucket_batch_sizes`, while PT `BucketBatcher` only has a single `batch_size` argument. So `BucketBatcher` is less powerful here. I wonder whether a single `batch_size` really makes sense. For buckets with shorter sequences, you could have a larger batch size than for those buckets with longer sequences.
